File: daemon/isoinfo.c
/* libguestfs - the guestfsd daemon
 * Copyright (C) 2012 Red Hat Inc.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

#include <config.h>

#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <inttypes.h>
#include <string.h>
#include <unistd.h>
#include <time.h>

#include "guestfs_protocol.h"
#include "daemon.h"
#include "actions.h"

static int
parse_uint32 (uint32_t *ret, const char *str)
{
  uint32_t r;

  if (sscanf (str, "%" SCNu32, &r) != 1) {
    reply_with_error ("cannot parse numeric field from isoinfo: %s", str);
    return -1;
  }

  *ret = r;
  return 0;
}

/* This is always in a fixed format:
 * "2012 03 16 11:05:46.00"
 * or if the field is not present, then:
 * "0000 00 00 00:00:00.00"
 */
static int
parse_time_t (int64_t *ret, const char *str)
{
  struct tm tm;
  time_t r;

  if (STREQ (str, "0000 00 00 00:00:00.00")) {
    *ret = -1;
    return 0;
  }

  if (sscanf (str, "%04d %02d %02d %02d:%02d:%02d",
              &tm.tm_year, &tm.tm_mon, &tm.tm_mday,
              &tm.tm_hour, &tm.tm_min, &tm.tm_sec) != 6) {
    reply_with_error ("cannot parse date from isoinfo: %s", str);
    return -1;
  }

  /* Adjust fields. */
  tm.tm_year -= 1900;
  tm.tm_mon--;
  tm.tm_isdst = -1;

  /* Convert to time_t. */
  r = timegm (&tm);
  if (r == -1) {
    reply_with_error ("invalid date or time: %s", str);
    return -1;
  }

  *ret = r;
  return 0;
}

static guestfs_int_isoinfo *
parse_isoinfo (char **lines)
{
  guestfs_int_isoinfo *ret;
  size_t i;

  ret = calloc (1, sizeof *ret);
  if (ret == NULL) {
    reply_with_perror ("calloc");
    return NULL;
  }

  /* Default each int field in the struct to -1. */
  ret->iso_volume_space_size = (uint32_t) -1;
  ret->iso_volume_set_size = (uint32_t) -1;
  ret->iso_volume_sequence_number = (uint32_t) -1;
  ret->iso_logical_block_size = (uint32_t) -1;
  ret->iso_volume_creation_t = -1;
  ret->iso_volume_modification_t = -1;
  ret->iso_volume_expiration_t = -1;
  ret->iso_volume_effective_t = -1;

  for (i = 0; lines[i] != NULL; ++i) {
    if (STRPREFIX (lines[i], "System id: ")) {
      ret->iso_system_id = strdup (&lines[i][11]);
      if (ret->iso_system_id == NULL) goto error;
    }
    else if (STRPREFIX (lines[i], "Volume id: ")) {
      ret->iso_volume_id = strdup (&lines[i][11]);
      if (ret->iso_volume_id == NULL) goto error;
    }
    else if (STRPREFIX (lines[i], "Volume set id: ")) {
      ret->iso_volume_set_id = strdup (&lines[i][15]);
      if (ret->iso_volume_set_id == NULL) goto error;
    }
    else if (STRPREFIX (lines[i], "Publisher id: ")) {
      ret->iso_publisher_id = strdup (&lines[i][14]);
      if (ret->iso_publisher_id == NULL) goto error;
    }
    else if (STRPREFIX (lines[i], "Data preparer id: ")) {
      ret->iso_data_preparer_id = strdup (&lines[i][18]);
      if (ret->iso_data_preparer_id == NULL) goto error;
    }
    else if (STRPREFIX (lines[i], "Application id: ")) {
      ret->iso_application_id = strdup (&lines[i][16]);
      if (ret->iso_application_id == NULL) goto error;
    }
    else if (STRPREFIX (lines[i], "Copyright File id: ")) {
      ret->iso_copyright_file_id = strdup (&lines[i][19]);
      if (ret->iso_copyright_file_id == NULL) goto error;
    }
    else if (STRPREFIX (lines[i], "Abstract File id: ")) {
      ret->iso_abstract_file_id = strdup (&lines[i][18]);
      if (ret->iso_abstract_file_id == NULL) goto error;
    }
    else if (STRPREFIX (lines[i], "Bibliographic File id: ")) {
      ret->iso_bibliographic_file_id = strdup (&lines[i][23]);
      if (ret->iso_bibliographic_file_id == NULL) goto error;
    }
    else if (STRPREFIX (lines[i], "Volume size is: ")) {
      if (parse_uint32 (&ret->iso_volume_space_size, &lines[i][16]) == -1)
        goto error;
    }
    else if (STRPREFIX (lines[i], "Volume set size is: ")) {
      if (parse_uint32 (&ret->iso_volume_set_size, &lines[i][20]) == -1)
        goto error;
    }
    else if (STRPREFIX (lines[i], "Volume set sequence number is: ")) {
      if (parse_uint32 (&ret->iso_volume_sequence_number, &lines[i][31]) == -1)
        goto error;
    }
    else if (STRPREFIX (lines[i], "Logical block size is: ")) {
      if (parse_uint32 (&ret->iso_logical_block_size, &lines[i][23]) == -1)
        goto error;
    }
    else if (STRPREFIX (lines[i], "Creation Date:     ")) {
      if (parse_time_t (&ret->iso_volume_creation_t, &lines[i][19]) == -1)
        goto error;
    }
    else if (STRPREFIX (lines[i], "Modification Date: ")) {
      if (parse_time_t (&ret->iso_volume_modification_t, &lines[i][19]) == -1)
        goto error;
    }
    else if (STRPREFIX (lines[i], "Expiration Date:   ")) {
      if (parse_time_t (&ret->iso_volume_expiration_t, &lines[i][19]) == -1)
        goto error;
    }
    else if (STRPREFIX (lines[i], "Effective Date:    ")) {
      if (parse_time_t (&ret->iso_volume_effective_t, &lines[i][19]) == -1)
        goto error;
    }
  }

  /* Any string fields which were not set above will be NULL.  However
   * we cannot return NULL fields in structs, so we convert these to
   * empty strings here.
   */
  if (ret->iso_system_id == NULL) {
    ret->iso_system_id = strdup ("");
    if (ret->iso_system_id == NULL) goto error;
  }
  if (ret->iso_volume_id == NULL) {
    ret->iso_volume_id = strdup ("");
    if (ret->iso_volume_id == NULL) goto error;
  }
  if (ret->iso_volume_set_id == NULL) {
    ret->iso_volume_set_id = strdup ("");
    if (ret->iso_volume_set_id == NULL) goto error;
  }
  if (ret->iso_publisher_id == NULL) {
    ret->iso_publisher_id = strdup ("");
    if (ret->iso_publisher_id == NULL) goto error;
  }
  if (ret->iso_data_preparer_id == NULL) {
    ret->iso_data_preparer_id = strdup ("");
    if (ret->iso_data_preparer_id == NULL) goto error;
  }
  if (ret->iso_application_id == NULL) {
    ret->iso_application_id = strdup ("");
    if (ret->iso_application_id == NULL) goto error;
  }
  if (ret->iso_copyright_file_id == NULL) {
    ret->iso_copyright_file_id = strdup ("");
    if (ret->iso_copyright_file_id == NULL) goto error;
  }
  if (ret->iso_abstract_file_id == NULL) {
    ret->iso_abstract_file_id = strdup ("");
    if (ret->iso_abstract_file_id == NULL) goto error;
  }
  if (ret->iso_bibliographic_file_id == NULL) {
    ret->iso_bibliographic_file_id = strdup ("");
    if (ret->iso_bibliographic_file_id == NULL) goto error;
  }

  return ret;

 error:
  free (ret->iso_system_id);
  free (ret->iso_volume_id);
  free (ret->iso_volume_set_id);
  free (ret->iso_publisher_id);
  free (ret->iso_data_preparer_id);
  free (ret->iso_application_id);
  free (ret->iso_copyright_file_id);
  free (ret->iso_abstract_file_id);
  free (ret->iso_bibliographic_file_id);
  free (ret);
  return NULL;
}

static guestfs_int_isoinfo *
isoinfo (const char *path)
{
  char *out = NULL, *err = NULL;
  int r;
  char **lines = NULL;
  guestfs_int_isoinfo *ret = NULL;

  /* --debug is necessary to get additional fields, in particular
   * the date & time fields.
   */
  r = command (&out, &err, "isoinfo", "--debug", "-d", "-i", path, NULL);
  if (r == -1) {
    reply_with_error ("%s", err);
    goto done;
  }

  lines = split_lines (out);
  if (lines == NULL)
    goto done;

  ret = parse_isoinfo (lines);
  if (ret == NULL)
    goto done;

 done:
  free (out);
  free (err);
  if (lines)
    free_strings (lines);

  return ret;
}

guestfs_int_isoinfo *
do_isoinfo_device (const char *device)
{
  return isoinfo (device);
}

guestfs_int_isoinfo *
do_isoinfo (const char *path)
{
  char *buf;
  guestfs_int_isoinfo *ret;

  buf = sysroot_path (path);
  if (!buf) {
    reply_with_perror ("malloc");
    return NULL;
  }

  ret = isoinfo (buf);
  free (buf);

  return ret;
}

File: daemon/isoinfo.ml
(* Parse isoinfo or xorriso output.
 * Copyright (C) 2009-2023 Red Hat Inc.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 *)

open Printf
open Unix

open Std_utils

include Structs

(* We treat ISO "strA" and "strD" formats the same way, simply
 * discarding any trailing spaces.
 *)
let iso_parse_strA str =
  let len = String.length str in
  let rec loop len =
    if len > 0 && str.[len-1] = ' ' then loop (len-1) else len
  in
  let len = loop len in
  String.sub str 0 len

let iso_parse_strD = iso_parse_strA

(* These always parse the intX_LSB (little endian) version. *)
let iso_parse_int16 s = s |> int_of_le16 |> Int64.to_int32
let iso_parse_int32 s = s |> int_of_le32 |> Int64.to_int32

(* Parse ISO dec-datetime to a Unix time_t. *)
let iso_parse_datetime str =
  if String.sub str 0 16 = "0000000000000000" then -1_L
  else (
    let tm_year = int_of_string (String.sub str 0 4) in
    let tm_mon = int_of_string (String.sub str 4 2) in
    let tm_mday = int_of_string (String.sub str 6 2) in
    let tm_hour = int_of_string (String.sub str 8 2) in
    let tm_min = int_of_string (String.sub str 10 2) in
    let tm_sec = int_of_string (String.sub str 12 2) in

    (* Adjust fields. *)
    let tm_year = tm_year - 1900 in
    let tm_mon = tm_mon - 1 in

    (* Convert to time_t in UTC timezone. *)
    let tm = { tm_sec; tm_min; tm_hour; tm_mday; tm_mon; tm_year;
               tm_wday = -1; tm_yday = -1; tm_isdst = false } in
    let old_TZ = try Some (getenv "TZ") with Not_found -> None in
    putenv "TZ" "UTC";
    let r = Int64.of_float (fst (mktime tm)) in
    Option.iter (putenv "TZ") old_TZ;

    (* The final byte is a time zone offset from GMT.
     *
     * The documentation of this at
     * https://wiki.osdev.org/ISO_9660#The_Primary_Volume_Descriptor
     * is wrong.  See the ECMA 119 documentation for a correct
     * description.
     *
     * For a disk image which we know was created
     * in BST (GMT+1), this contains 0x4, ie. 4 * 15 mins ahead.
     * We have to subtract this from the gmtime above.
     *)
    let tz = Char.code str.[16] in
    let tz = if tz >= 128 then tz - 256 else tz in
    r -^ (Int64.of_int (tz * 15 * 60))
  )

let isoinfo_device dev =
  let r, pvd =
    with_openfile dev [O_RDONLY] 0 (
        fun fd ->
        ignore (lseek fd 32768 SEEK_SET);
        let pvd = Bytes.create 2048 in
        let r = read fd pvd 0 2048 in
        r, Bytes.to_string pvd
      ) in

  let sub = String.sub pvd in

  (* Check that it looks like an ISO9660 Primary Volume Descriptor.
   * https://wiki.osdev.org/ISO_9660#The_Primary_Volume_Descriptor
   *)
  if r <> 2048 || pvd.[0] <> '\001' || sub 1 5 <> "CD001" then
    failwithf "%s: not an ISO file or CD-ROM" dev;

  (* Parse out the PVD fields. *)
  let iso_system_id = sub 8 32 |> iso_parse_strA in
  let iso_volume_id = sub 40 32 |> iso_parse_strA in
  let iso_volume_space_size = sub 80 4 |> iso_parse_int32 in
  let iso_volume_set_size = sub 120 2 |> iso_parse_int16 in
  let iso_volume_sequence_number = sub 124 2 |> iso_parse_int16 in
  let iso_logical_block_size = sub 128 2 |> iso_parse_int16 in
  let iso_volume_set_id = sub 190 128 |> iso_parse_strD in
  let iso_publisher_id = sub 318 128 |> iso_parse_strA in
  let iso_data_preparer_id = sub 446 128 |> iso_parse_strA in
  let iso_application_id = sub 574 128 |> iso_parse_strA in
  let iso_copyright_file_id = sub 702 37 |> iso_parse_strD in
  let iso_abstract_file_id = sub 739 37 |> iso_parse_strD in
  let iso_bibliographic_file_id = sub 776 37 |> iso_parse_strD in
  let iso_volume_creation_t = sub 813 17 |> iso_parse_datetime in
  let iso_volume_modification_t = sub 830 17 |> iso_parse_datetime in
  let iso_volume_expiration_t = sub 847 17 |> iso_parse_datetime in
  let iso_volume_effective_t = sub 864 17 |> iso_parse_datetime in

  (* Return the struct. *)
  {
    iso_system_id; iso_volume_id; iso_volume_space_size;
    iso_volume_set_size; iso_volume_sequence_number;
    iso_logical_block_size; iso_volume_set_id; iso_publisher_id;
    iso_data_preparer_id; iso_application_id; iso_copyright_file_id;
    iso_abstract_file_id; iso_bibliographic_file_id;
    iso_volume_creation_t; iso_volume_modification_t;
    iso_volume_expiration_t; iso_volume_effective_t;
  }

let isoinfo file =
  let chroot = Chroot.create ~name:(sprintf "isoinfo: %s" file) () in
  Chroot.f chroot isoinfo_device file
